File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int tc = Integer.parseInt(br.readLine());

        int N, M;
        LinkedList<int[]> queue;
        for(int i = 0; i < tc; i++){
            st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken()); // 문서의 갯수
            M = Integer.parseInt(st.nextToken()); // 몇번째로 인쇄되었는지 궁금한 문서의 queue index

            st = new StringTokenizer(br.readLine());
            queue = new LinkedList<>();

            int cnt = 0; // 인쇄 횟수
            for(int j = 0; j < N; j++){ // queue에 {문서번호, 중요도} 추가
                queue.addLast(new int[]{j, Integer.parseInt(st.nextToken())});
            }

            while(!queue.isEmpty()){

                int[] first = queue.removeFirst(); // 가장 앞에 있는 원소
                boolean isMax = true;

                for(int k = 0; k < queue.size(); k++){ // first가 최대 중요도값을 가지는지 체크를 위한 iteration

                    if(first[1] < queue.get(k)[1]){ // 가장 앞에 있는 first가 max가 아닐 경우
                        queue.addLast(first); // first와 k인덱스 이전의 원소들을 차례로 뒤로 보낸다.

                        for(int l = 0; l < k; l++){
                            queue.addLast(queue.removeFirst());
                        }

                        isMax = false;
                        break;
                    }
                }

                if(!isMax){ // 최대 중요도값을 가지지 않으면 다시 반복
                    continue;
                }

                // 최대 중요도 값을 지니는 것이었으면 cnt++
                cnt++;
                if(first[0] == M){ // 원하는 문서가 인쇄가 되면 break
                    break;
                }
            }
            sb.append(cnt).append("\n");
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/1966. 프린터 큐/프린터 큐.kt
import java.io.*
import java.util.*
import kotlin.collections.ArrayDeque

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    class Data(val index : Int, val weight : Int)
    val t = readLine().toInt()
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    repeat(t) {
        var st = StringTokenizer(readLine())
        val n = st.nextToken().toInt() // 문서의 갯수
        val m = st.nextToken().toInt() // 목표
        var cnt = 0 // 문서 출력 순서

        val queue = ArrayDeque<Data>()

        st = StringTokenizer(readLine())
        for(i in 0 until n) {
            val weight = st.nextToken().toInt() // 문서의 중요도
            queue.add(Data(i, weight))
        }

        while(!queue.isEmpty()) {
            val removed = queue.removeFirst()
            var maxWeight = true

            for(i in 0 until queue.size) {
                // queue를 돌면서 현재 문서의 중요도가 제일 높은지 체크
                if(removed.weight < queue[i].weight) {
                    queue.addLast(removed)
                    maxWeight = false

                    for(j in 0 until i) {
                        queue.addLast(queue.removeFirst())
                    }
                    break
                }
            }

            if(maxWeight) {
                cnt++
                // 제거된 문서의 순서가 m과 같으면 write
                if(removed.index == m) {
                    bw.write("$cnt\n")
                    break
                }
            }
        }
    }
    bw.flush()
    bw.close()
}